### PR TITLE
[WPE][GTK] Handle the flakiness in test /webkit/Downloads/local-file-error.

### PR DESCRIPTION
**Pull request by ziransun**
#### ea82aedda965972b25919d22dd960ded8f54e66f
<pre>
[WPE][GTK] Handle the flakiness in test /webkit/Downloads/local-file-error.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281813">https://bugs.webkit.org/show_bug.cgi?id=281813</a>

Reviewed by Michael Catanzaro.

In one sub-test in /webkit/Downloads/local-file-error, it downloads
a local file and immediately cancel it.

The flakiness is caused by how fast the download finishes. There is a chance
that the DownloadManager has called downloadFinished() before calling
cancelDownload(), in which case the download with downloadID has been
removed frome m_downloads. This causes ASSERT_NOT_REACHED() failure. Yet in
DownloadManager we are not able to get information about cancellation until
cancelDownload() call. The assertion is removed to address this case.

The solution here is, keeping a track of the cancellation status in DownloadProxy,
and handling cancellation case from there. This will fix sequent assertion
failures after the downloadManager calls.

* Tools/TestWebKitAPI/glib/TestExpectations.json:
* Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:
(WebKit::DownloadManager::cancelDownload):
* Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:
(WebKit::DownloadProxy::cancel):
(WebKit::DownloadProxy::didFinish):
(WebKit::DownloadProxy::didFail):
* Source/WebKit/UIProcess/Downloads/DownloadProxy.h:

Canonical link: <a href="https://commits.webkit.org/287675@main">https://commits.webkit.org/287675@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/65e93ec672750e25a884f4a8b2f6fe05d8f2b35d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59475 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84989 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31450 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/82579 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7778 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62878 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20683 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83537 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73262 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43181 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/50273 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27421 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29909 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71417 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27939 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86423 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7694 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5436 "Found 1 new test failure: imported/w3c/web-platform-tests/dom/observable/tentative/observable-filter.any.worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71168 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7869 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69100 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70406 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17537 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14409 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13355 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7656 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13175 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7495 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11014 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9300 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp [hunk 1]
```diff
@@ -127,7 +127,10 @@ void DownloadManager::cancelDownload(DownloadID downloadID, CompletionHandler<vo
         pendingDownload->cancel(WTFMove(completionHandler));
         return;
     }
+    // There is a chance that the DownloadManager has called downloadFinished() and the download with downloadID has been removed frome m_downloads.
+#if !PLATFORM(GTK) && !PLATFORM(WPE)
     ASSERT_NOT_REACHED();
+# endif
```
Comment (mcatanzaro) [inline Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp]: Please don't make the behavior platform-specific! You can just remove the assert.

File: Source/WebKit/UIProcess/API/glib/WebKitDownloadClient.cpp [hunk 2]
```diff
@@ -103,8 +103,12 @@ class DownloadClient final : public API::DownloadClient {
 
     void didFinish(DownloadProxy& downloadProxy) override
     {
-        if (webkitDownloadIsCancelled(m_download.get()))
+        if (webkitDownloadIsCancelled(m_download.get())) {
+#if PLATFORM(GTK) || PLATFORM(WPE)
+            downloadProxy.setDownloadIsCancelled(true);
+#endif
             return;
+        }
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/API/glib/WebKitDownloadClient.cpp]: Is this change needed? I think if webkitDownloadIsCancelled returns true here, then DownloadProxy::cancel should have already been called, and there you already call setDownloadIsCancelled.

File: Source/WebKit/UIProcess/Downloads/DownloadProxy.h [hunk 3]
```diff
@@ -104,6 +104,10 @@ class DownloadProxy : public API::ObjectImpl<API::Object::Type::Download>, publi
     void setProgress(NSProgress *progress) { m_progress = progress; }
     NSProgress *progress() const { return m_progress.get(); }
 #endif
+#if PLATFORM(GTK) || PLATFORM(WPE)
+    void setDownloadIsCancelled(bool cancelled) { m_downloadIsCancelled = cancelled; }
+    bool downloadIsCancelled() const { return m_downloadIsCancelled; }
+#endif
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Downloads/DownloadProxy.h]: So the first thing I don't like is that these are platform-specific. It's better to compile this on all platforms even if it's only used by GTK/WPE, so I would remove all the PLATFORM() guards.

Next, we don't actually need the downloadIsCancelled() member function. That can be replaced by checking m_downloadIsCancelled directly.

Then I would suggest renaming `setDownloadIsCancelled` to simply `cancel`. You never need to un-cancel a download, so there's no need for the bool parameter. That can be removed.

This leaves one last problem: after my suggested changes, it's redundant with the existing cancel function! I think you can get rid of your new function and just write `m_downloadIsCancelled = true` at the top of DownloadProxy::cancel.

**Review by mcatanzaro — CHANGES_REQUESTED** [review]
Thanks for fixing these API tests.

**Comment by ziransun** [comment]
Thanks very much for the review comments! Code is now updated. PTAL. Thank you!

File: Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp [hunk 4]
```diff
@@ -127,7 +127,6 @@ void DownloadManager::cancelDownload(DownloadID downloadID, CompletionHandler<vo
         pendingDownload->cancel(WTFMove(completionHandler));
         return;
     }
-    ASSERT_NOT_REACHED();
```
Comment (mcatanzaro) [inline Source/WebKit/NetworkProcess/Downloads/DownloadManager.cpp:130]: Might be wise to add a short comment here:

```
// If there is no active or pending download, then the download finished in a short race window after cancellation was requested.
```

File: Source/WebKit/UIProcess/Downloads/DownloadProxy.h [hunk 5]
```diff
@@ -165,6 +165,7 @@ class DownloadProxy : public API::ObjectImpl<API::Object::Type::Download>, publi
 #if HAVE(MODERN_DOWNLOADPROGRESS)
     RefPtr<ProcessAssertion> m_assertion;
 #endif
+    bool m_downloadIsCancelled { false };
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Downloads/DownloadProxy.h]: You can optimize the struct layout slightly by moving this beneath `m_wasUserInitiated`.

File: Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp [hunk 6]
```diff
@@ -224,6 +225,8 @@ void DownloadProxy::didFinish()
     updateQuarantinePropertiesIfPossible();
 #endif
     m_client->didFinish(*this);
+    if (m_downloadIsCancelled)
+        return;
```
Comment (mcatanzaro) [inline Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:229]: It looks like the intended behavior here is that either `DownloadClient::didFinish` or `DownloadClient::didFail` gets called exactly once. But I'm not sure whether canceling triggers didFail to be called or not? Do you know if it does?

 * If it doesn't, then your change here is good.
 * But if it does, then we'll wind up with didFail being called and then didFinish also being called next, which is not desired. In this case, we should check `if (m_downloadIsCancelled)` immediately _before_ `m_client->didFinish` rather than after. And same change would be needed in `DownloadProxy::didFail` as well.
 
 Please check!

**Review by mcatanzaro — COMMENTED** [review]
OK, this looks much better.

I think you should also check `if (m_downloadIsCancelled)` in `DownloadProxy::didFail` as well, since it's going to have the same problem.

File: Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp (hunk 6 above)
Comment (ziransun) [reply Source/WebKit/UIProcess/Downloads/DownloadProxy.cpp:229]: It doesn't seem triggering didFail.

**Review by mcatanzaro — APPROVED** [review]
